### ISSUE TITLE
Multi-session debug adapters (nodejs support)

### ISSUE THREAD

**Issue by entropitor** [mfussenegger/nvim-dap#136]
https://microsoft.github.io/debug-adapter-protocol/overview#debug-session-start

The DAP protocol supports multi-sessions (and e.g. JS debugger seems to be using that: https://github.com/microsoft/vscode-js-debug/issues/902), does this plugin support this somehow? It's okay if there needs to be an extension as for python but I'm wondering where to start.

**Comment by entropitor** [comment]
I tried using a callback for the server configuration but it doesn't seem to actually work yet

```lua
local M = {}
local uv = vim.loop

function M.__reload()
    package.loaded["dap.ext.js"] = nil
    print("Reloaded dap.ext.js")
end

local stdin = uv.new_pipe(false)
local stdout = uv.new_pipe(false)
local stderr = uv.new_pipe(false)
local handle = nil
function M.stop()
    print("stopping")

    stdin:close()
    stdin = uv.new_pipe(false)

    stdout:close()
    stdin = uv.new_pipe(false)

    stderr:close()
    stderr = uv.new_pipe(false)

    handle:close()
    handle = nil
end

function M.connect(callback, _config)
    if handle ~= nil then
        M.stop()
    end

    local pid_or_err
    handle, pid_or_err =
        uv.spawn(
        "node",
        {
            args = {
                os.getenv("HOME") .. "/.vim/bundle/vscode-js-debug/out/src/vsDebugServer.js"
            },
            stdio = {stdin, stdout, stderr},
            detached = true
        },
        function()
            M.stop()
            print("exited")
        end
    )
    print(vim.inspect({handle, pid_or_err}))
    stdout:read_start(
        function(_err, chunk)
            local port = chunk:gsub("\n", "")
            print(port)
            callback(
                {
                    type = "server",
                    port = port
                }
            )
        end
    )
end

return M

```

**Comment by entropitor** [comment]
It looks like things are gonna be a bit more complex: https://github.com/microsoft/vscode-js-debug/issues/902#issuecomment-814655262

It feels like we're gonna need to have the concept of multiple sessions in nvim-dap and/or a way to preprocess all messages we send so we can attach a sessionId

Also, we might have to postpone sending the breakpoints as the rootsession doesn't accept it

**Comment by entropitor** [comment]
Btw this is a follow-up to https://github.com/mfussenegger/nvim-dap/issues/82

**Comment by entropitor** [comment]
Current script:

```
local M = {}
local uv = vim.loop

function M.__reload()
    package.loaded["dap.ext.js"] = nil
    print("Reloaded dap.ext.js")
end

local onexit = nil
function M.stop()
    print("stopping")
    if onexit ~= nil then
        print("exiting")
        onexit()
    end
end

function M.connect(callback, _config)
    M.stop()

    local stdin = uv.new_pipe(false)
    local stdout = uv.new_pipe(false)
    local stderr = uv.new_pipe(false)
    local handle, pid_or_err
    onexit = function()
        print("stopping")
        stdin:close()
        stdout:close()
        stderr:close()
        handle:close()
        handle:kill(9)
    end
    handle, pid_or_err =
        uv.spawn(
        "node",
        {
            args = {
                os.getenv("HOME") .. "/.vim/bundle/vscode-js-debug/out/src/vsDebugServer.js"
            },
            stdio = {stdin, stdout, stderr},
            detached = true
        },
        function()
            print("spawn.onExit called")
            M.stop()
        end
    )
    assert(handle, "Error trying to get DAP pid: " .. pid_or_err)
    print(vim.inspect({handle, pid_or_err}))
    stdout:read_start(
        function(_err, chunk)
            local port = chunk:gsub("\n", "")
            print(port)
            callback(
                {
                    type = "server",
                    port = port
                }
            )
        end
    )
end

vim.api.nvim_command("autocmd VimLeavePre * lua require('dap.ext.js').stop()")

return M
```

and
```
dap.adapters["pwa-node"] = require("dap.ext.js").connect
dap.configurations.typescript = {
    {
        type = "pwa-node",
        request = "attach",
        name = "Attach",
        continueOnAttach = true,
        attachExistingChildren = true,
        port = 9229
    },
    {
        type = "pwa-node",
        request = "launch",
        name = "Run jest",
        localRoot = "${workspaceFolder}",
        runtimeArgs = {"--inspect"},
        program = "./node_modules/.bin/jest",
        args = {
            "--no-coverage",
            "--",
            "src/index.spec.ts"
        }
    },
}
```

(This should clean up the start process better + using `pwa-node` instead of `node-terminal`, now I'm stuck on the custom message we need to implement, for which we might need multiple sessions (instead of singleton, or a way to preprocess a request before sending them so we can add the sessionId)

**Comment by mfussenegger** [comment]
See my comment [here](https://github.com/mfussenegger/nvim-dap/issues/82#issuecomment-761814861). Opposed to what they claim, vscode-js-debug is not DAP compliant as it doesn't work without their undocumented session-id extension. I won't be adding out-of-spec things to the core of nvim-dap.
The only thing I can imagine adding is some generic intercept point for requests and responses, so that they can be extended. That may already be possible in part with the `dap.listeners.{before | after}` hooks.


I suggested back than that they should open up an issue within [debug-adapter-protocol](https://github.com/microsoft/debug-adapter-protocol) to get their extension documented and added to the official specification. Unfortunately that didn't happen so far.

**Comment by entropitor** [comment]
I might work on it a bit more. I think we'll need a way to hook into requests, which should be relatively easy, the biggest problem is gonna be that the breakpoints are currently sent too early (before we have a sessionId) and so we'll have to postpone this somehow 😭

**Comment by mfussenegger** [comment]
>  I think we'll need a way to hook into requests, which should be relatively easy, the biggest problem is gonna be that the breakpoints are currently sent too early

It might be possible to circumvent that by doing the special "session-attach" request within a `listeners.before.event_initialized` hook.
If this request gets the response with the sessionId before the other callback is fired it should work out.

**Comment by entropitor** [comment]
If I'm understanding the answers on the other thread correctly, you need to send an attach on the root session, which will give you an answer in a custom message and you need to use an attach/launch config from that answer and use it in a new (child) session. So this would mean we would need more than 1 DAP session (a tree) while currently it's a singleton. I can try playing around with it as I get more answers. 

Would you be willing to accept a PR that can just handle more than 1 session without any protocol specific stuff (I get that that should live elsewhere)?

**Comment by theHamsta** [comment]
Would really enjoy having more than one session. It certainly involves some changes to get away from that single session global

**Comment by mfussenegger** [comment]
> Would you be willing to accept a PR that can just handle more than 1 session without any protocol specific stuff

Depends.  I think I'd need to see some high level sketch first to see in which direction it is going.

If it's a small change or opens up other use cases with other adapters that follow the spec - I guess yes

If it's too invasive and brings no benefit for other use cases I'd say no. In that case a dedicated dap proxy app that emulates a single session might be a workaround - one that could then also be used with other dap clients. 

> Would really enjoy having more than one session

Can you elaborate what you'd do with that?

**Comment by theHamsta** [comment]
I agree that more than one session complicates things for UI (also for the user). But there are some part of `dap.lua` that are unnecessarily using on local variables at global scope. I think it would be possible re-organize those few spots and move the those variables into the session object.

I think this would make the code a bit nicer and easier to understand and a bit more self-contained. When I saw the code for the first time I was a bit confused which code was using state from save in `self` and what was saved in global scope. When all state needed for one session in saved in the Session object, then users could also write configuration to start multiple of them.

One use-case I could think of would be that you leave one session paused in state where the bug occurs (maybe it's a rare heisebug) and then check whether you can reproduce it in simpler manner or understand how you even came to that point. Could have one "current session" `dap.session()` and you could switch between the active session next_session/previous_session.

You can already have multiple session by launching multiple nvim instances. So for me this would rather be an internal cleanup. But if it helps someone, why not allow it? Most parts are already per-session (requests, variables, frames, also event handlers). nvim-dap could still continue use its "current session" to keep the logic simple. Users could go multi-session in their configs by just using the public methods of the Session object.

**Comment by mfussenegger** [comment]
> But there are some part of dap.lua that are unnecessarily using on local variables at global scope

Not sure if there is really much left that is unnecessarily using local variables at global scope. Looking through it:

- Breakpoint information is stored globally because they can be created before a session is started and survive across sessions. This is one aspect that gets tricky if we introduce multiple parallel sessions - how would we map breakpoints to sessions.

- `last_run` information - this is by design required to survive sessions.
- `terminal_buf` debatable - currently the buffer survives until after the session - which is useful because you may still want to read the output. There is now only a single terminal possible, so starting a new session clears it. This was done to improve the UX to avoid cluttering things with multiple terminals.


Then there are some configuration aspects, some of them are clearly global only, for some you could argue that it might make sense to be able to override them per session and have the globals only as fall-back, similar to the LSP handler model:

- `configurations`
- `adapters`
- `defaults`
- `listeners`


If you use `attach` or `launch` directly you can already get multiple concurrent sessions - the problem are mostly the side effects of the session running and interacting with the UI - E.g. jumping to the stopped breakpoints and so on. This is I think the biggest challenge. It complicates the UI and complicates the cognitive model for the user. Not sure if it's possible to support it without burdening users who *don't need it* with some additional decisions they have to make.

(I think we're also mixing a general "user can have multiple sessions in parallel" with the nodejs specific "there is a root-session and child sessions" requirement, which is similar but not necessarily the same thing.) 


> But if it helps someone, why not allow it?

Because that can quickly lead to feature creep :)


> Users could go multi-session in their configs by just using the public methods of the Session object.

As mentioned, by using `attach` or `launch` directly they could already do something like that, the methods returns the session instance. Maybe we'd have to decouple the `Session:spawn` or `Session:connect` call - or make these accessible directly - and then the flow necessariy to get nodejs could be implemented.

**Comment by theHamsta** [comment]
> As mentioned, by using `attach` or `launch` directly they could already do something like that, the methods returns the session instance.

This will not work at the moment as some parts of the code base refer to `self` and other to `session`. So as soon as you would have two session they would randomly mutate the other object. Making this work is the only thing I would like to achieve.

**Comment by mfussenegger** [comment]
> > As mentioned, by using `attach` or `launch` directly they could already do something like that, the methods returns the session instance.
> 
> This will not work at the moment as some parts of the code base refer to `self` and other to `session`. So as soon as you would have two sessions they would randomly mutate either self or the other object. Making this work is the only thing I would like to achieve. This is what I meant with clean up: just to have let session behave like a real object.

Ah, yes that makes sense. I misunderstood you before. 
I think extracting the Session stuff into a separate file as done in TJs PR will help with this, because `session` won't be in scope anymore. 

It's still on my list to wrap that up

**Comment by entropitor** [comment]
> Breakpoint information is stored globally because they can be created before a session is started and survive across sessions. This is one aspect that gets tricky if we introduce multiple parallel sessions - how would we map breakpoints to sessions.

From what I gather from the vscode-js-debug thread, vscode sends breakpoints to ALL sessions, (which also means that if they get rejected, they don't get cleared)

**Comment by mfussenegger** [comment]
Just had a closer look to see what would be required.

The adapter definition and the configuration can be simplified to:


```lua
dap.adapters['pwa-node'] = {
  type = 'executable',
  command = 'node',
  args = {HOME .. '/dev/microsoft/vscode-js-debug/out/src/flatSessionLauncher.js'},
}
dap.configurations.javascript = {
  {
    type = 'pwa-node',
    request = 'launch',
    name = 'Launch program',
    program = '${file}'
  },
}
```


Then I think it would be necessary to add support for reverse request custom listeners. See https://github.com/mfussenegger/nvim-dap/pull/145/files

Then one could add something like this:

```lua
dap.listeners.after['reverse_request_attachedChildSession']['js'] = function(session, request)
  print(vim.inspect(request))

  -- TODO: 
  -- - call dap.attach or dap.launch with the config, depending on config.type
  -- - manage the sessions somehow, `session` is the root-session, whatever `attach`/`launch` return is the child session
  -- - Would need to override session.request to inject sessionId to each request
  -- 
  --  Open question:
  --    Do further actions need to use the root session, or should the child session be used? Fine to use all?
  --
end
```

**Comment by entropitor** [comment]
@mfussenegger I had something similar but I used the vsDebugServer. In that case the "child session" is just a new session on a new port and we need to make sure not to stop the root session, but afterwards, we don't need to do any processing anymore (e.g. no need for adding a sessionId to every command).

Instead of a raw table for the adapter, I use the function-style that takes the config and if it's for a child session, it just connects to it, (it gets the port out of the config) and otherwise it starts the root sesion.

> Do further actions need to use the root session, or should the child session be used? Fine to use all?

They need to be send to the child session AFAIK, however, I wasn't able to get this fully working as the breakpoints still get rejected, at least for my typescript jest example I was trying to get working.

The breakpoints need to be send to all clients though. VSCode seems to set the breakpoint to "unbound" breakpoint when the server rejects it but doesn't delete it and when there is a new session, sends all breakpoints again. This seems a better implementation overall

In general, the server allows more than 1 child session but it's only if you use child processes e.g. jest workers, ... I think it would be awesome if we can get it to work with just 1 child session already. I'm not sure if we should immediately try to get all childs to work.

**Comment by mfussenegger** [comment]
Just to keep this up2date, there is an issue in vscode-js-debug to support the regular DAP flow: https://github.com/microsoft/vscode-js-debug/issues/969

**Comment by mfussenegger** [comment]
debugpy also uses  multi sessions.  One difference is that it uses an event to send the attach configuration. 

And the breakpoints apply to whatever session you send them to. I got an initial implementation here in case you're still interested in getting this to work with js-debug and want to have a look https://github.com/mfussenegger/nvim-dap-python/pull/21

**Comment by mfussenegger** [comment]
@entropitor do you intend to extract that into a dedicated plugin?

**Comment by entropitor** [comment]
I might at some point but I need to find some time to do it properly. I also didn't fully have it working, I only get jest working (and even then, it looks like the breakpoints get rejected). Not sure why it doesn't work but with being able to debug jest tests, I can manage for now. 

I haven't tried your idea for closing the root session properly but I can try that later and if it doesn't work, file a new issue.

**Comment by mxsdev** [comment]
So, I spent some time with this and got it to work fully with node.js on my machine. [Here's the plugin](https://github.com/mxsdev/nvim-dap-vscode-js) I made for it. For any interested, here's what I found:

- Since the debugger uses its own session system, verification of breakpoints must be done asynchronously. To accomplish that with the DAP system, the debugger responds to the `setBreakpoints` request initially with a status of false (rejected), and a nonsense message (`Unbound breakpoint`), and then sends a `breakpoint` event with a `reason` of `changed`, and that's where the verification status of the breakpoint is corrected.
- A child session is isomorphic to a root session, in that it can _itself_ spawn new child sessions. Allowing this behavior was necessary for jest to work on my system and ubuntu containers, even with the `--runInBand` flag set.

To solve the breakpoints issue in the plugin, I use the `dap.listeners.before` hook to set a js session's breakpoint status to verified by default, and then defer breakpoint rejection to the time of the first `continued` event, which the vscode debugger always sends when the debugger starts running.

I was thinking it might make sense for this to be implemented in this plugin, rather than mine. Probably what makes sense is to just change the relevant breakpoint if it exists upon receiving `event_breakpoint`, and if possible, log breakpoint rejection silently (or give an adapter flag to do so). Unfortunately, the DAP spec leaves it very ambiguous how the client is supposed to act in response to the `breakpoint` event, so it may be best to not implement it.

By the way, @entropitor, I was going to just say thank you for sharing your dotfiles, but I couldn't believe my eyes when I saw your repos; years ago, I used your work on the AutoReferee HUD client to build a new one. So, by some insane coincidence, here I am, years later, once again building off of some very obscure code you wrote, and I guess I have two things to thank you for :)
